Making DimShuffles view by default

Sentient07 opened the pull request:
and adding `_props_dict` to few Ops in old backend

Sentient07 commented:
@nouiz , I'm not sure if this is what the issue meant. From my understandings, I need to use the ```_props_dict()``` method for GPU ops to be consistent with the CPU Ops. I was able to do that for only one GPU OP. The rest of them are giving me various errors. I'll compile all the errors and add a comment. (Sorry for the delayed PR, there was a delay in shipping of my new notebook)

nouiz commented on theano/sandbox/cuda/opt.py:
```diff
@@ -370,7 +370,7 @@ def local_gpu_split(node):
             any(c != 'output' and isinstance(c.op, GpuFromHost) for c, idx
                 in outs_clients)):
-            new_op = GpuSplit(node.op.len_splits)
+            new_op = GpuSplit(node.op._props_dict()['len_splits'])
```
Here the ticket was asking that this was done:
GpuSplit(*node.op._props_dict())

This way, if a new props is done on the CPU, it will make the optimization crash if we don't update the GPU op. I'll kill travis as it is kind of busy and won't be useful for this commit now.

Sentient07 replied:
@nouiz ,  I will update the PR and will paste it if i have any errors. And yeah, understood the goal of the issue.

nouiz commented:
Can you give an example of error you got?

Sentient07 commented:
I had updated the ```node.op``` in ```GpuSplit``` op. 
But, when it try to use ```_props_dict()``` for ```GPUDimShuffle``` and ```GPUElemwise``` objects, I get an ```AttributeError``` that the objects have no attribute ```__props__```. 
The full traceback : 

```
ERROR (theano.gof.opt): Traceback (most recent call last):
  File "/Users/Ramana/projects/theano/Theano/theano/gof/opt.py", line 1779, in process_node
    replacements = lopt.transform(node)
  File "/Users/Ramana/projects/theano/Theano/theano/sandbox/cuda/opt.py", line 400, in local_gpu_dimshuffle_0
    new_op = GpuDimShuffle(*dimshuffle_node.op._props_dict())
  File "/Users/Ramana/projects/theano/Theano/theano/gof/op.py", line 812, in _props_dict
    for a in self.__props__])
AttributeError: 'DimShuffle' object has no attribute '__props__'

```

nouiz commented:
This is because the class DimShuffle don't implement the \__props__.

To implement it add this attribute to that class:

__props__ = ("input_broadcastable", "new_order", "inplace")

Also delete the method \__eq__

In a separate clean up commit, you can also delete the method \__getstate__
\__setstate__ \_rehash \__hash__ and the calls to \_rehash.

On Fri, Feb 12, 2016 at 6:54 AM, Ramana Subramanyam <
notifications@github.com> wrote:

> I had updated the node.op in GpuSplit op.
> But, when it try to use _props_dict() for GPUDimShuffle and GPUElemwise
> objects, I get an AttributeError that the objects have no attribute
> __props__.
> The full traceback :
>
> ERROR (theano.gof.opt): Traceback (most recent call last):
>   File "/Users/Ramana/projects/theano/Theano/theano/gof/opt.py", line 1779, in process_node
>     replacements = lopt.transform(node)
>   File "/Users/Ramana/projects/theano/Theano/theano/sandbox/cuda/opt.py", line 400, in local_gpu_dimshuffle_0
>     new_op = GpuDimShuffle(*dimshuffle_node.op._props_dict())
>   File "/Users/Ramana/projects/theano/Theano/theano/gof/op.py", line 812, in _props_dict
>     for a in self.__props__])
> AttributeError: 'DimShuffle' object has no attribute '__props__'
>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/4039#issuecomment-183294201>.
>

Sentient07 commented:
@nouiz yeah, wanted to confirm this. sure, will update

Sentient07 commented:
@nouiz , I've used ```_props_dict``` for all the GPU ops in ```theano/sandbox/cuda/opt.py``` . But, i'm not able to run the tests for ```theano/sandbox/gpuarray/tests/*.py ``` as i don't have CuDDN (my approval to join Accelerated Computing Developer Program hasn't been accepted yet). I'll update this PR once i get CuDNN

nouiz commented on theano/sandbox/cuda/basic_ops.py:
```diff
@@ -186,6 +186,8 @@ class GpuElemwise(GpuOp):
 
     """
 
+    __props__ = ("scalar_op", )
```
inplace_pattern is also a props.

nouiz replied:
sync too.

Delete the method \__eq__

Sentient07 replied:
@nouiz Will add inplace_pattern, but why delete that method ?

nouiz replied:
If \__props__ is used, it generate some default implementation of \__eq__ \__hash__ and \__str__. Sometimes we don't want to use the default method. So I just checked which one we want to remove and told you so.

Sentient07 replied:
Oh. Sure, would remove them.

nouiz commented on theano/sandbox/cuda/basic_ops.py:
```diff
@@ -569,6 +571,8 @@ class GpuCAReduce(GpuOp):
 
     """
 
+    __props__ = ("reduce_mask", "scalar_op", )
```
pre_scalar_op is also a props.

Also delete the method \__eq__ and \__hash__

nouiz commented on theano/tensor/elemwise.py:
```diff
@@ -492,6 +493,8 @@ class Elemwise(OpenMPOp):
 
     """
 
+    __props__ = ("scalar_op",)
```
add inplace_pattern

nouiz commented on theano/sandbox/cuda/opt.py:
```diff
@@ -1587,7 +1580,7 @@ def local_conv_gemm(node):
             kern = kern[:, :, ::-1, ::-1]
             # By default use GpuCorrMM
-            rval = GpuCorrMM(border_mode, subsample)(
+            rval = GpuCorrMM(**node.op._props_dict())(
```
Node is GpuConv and it don't implement \__props__ so this would fail if used.

GpuConv is hard to implement this \__props__, and is obsolete. So just don't change this line.

nouiz commented on theano/sandbox/cuda/basic_ops.py:
```diff
@@ -186,6 +186,8 @@ class GpuElemwise(GpuOp):
 
     """
 
+    __props__ = ("scalar_op", "inplace_pattern", )
```
I forgot, there should be "sync" in that list too.

Sentient07 replied:
Ah! Yeah, sorry, even i forgot.

nouiz commented on theano/sandbox/cuda/opt.py:
```diff
@@ -1165,11 +1162,9 @@ def local_gpu_advanced_incsubtensor1(node):
             if (compute_capability < 2 or
                 x.ndim != 2 or
                 y.ndim != 2):
-                gpu_op = GpuAdvancedIncSubtensor1(
```
you missed the one in line 1127 and 1130

nouiz commented on theano/tensor/elemwise.py:
```diff
@@ -125,6 +125,7 @@ class DimShuffle(Op):
 
     _f16_ok = True
     check_input = False
+    __props__ = ("input_broadcastable", "new_order")
```
I missed it, but inplace should be in this list. But this will cause problem on the GPU as it don't exist on GPU. The real solution would be to remove the inplace parameter and suppose it is always True. Then we will need to remove the opt dimshuffle_as_view and local_dimshuffle_no_inplace_at_canonicalize. Also, in the file tensor/basic.py, in the function transpose, remove the inplace parameter.

Maybe there is other clean up. So run the tests locally if possible.

thanks

Sentient07 replied:
Yes, i think i somewhat understand this. Will look into this and get back if i have any question regarding this. (Sorry for the delayed response, i was down with flu)

nouiz replied:
No problem. Take care of you. At the same time, this PR need a rebase when you have time to look at it again.

Sentient07 replied:
Thanks! :) I will rebase and address the remaining comment and push it before the weekend

Sentient07 replied:
@nouiz I don't think i can test it with GPU. There is an issue with GPU while running the tensor's tests. ( https://github.com/Theano/Theano/issues/4131#issuecomment-190013710). I've currently tested the tensor's test with CPU, and sandbox/cuda/tests/ with GPU. After making the changes suggested by you, two tests fail. The failure is after I remove the opt, ```dimshuffle_as_view```. I've updated the PR with the commit, please check travis for the error. I tried debugging the failure message, but unable to figure out how. 
@abergeron , @lamblin : Since @nouiz is on vacation, I would be really glad if you could have a look at this :)

lamblin replied:
The tests are not meant to be run with `device=gpu`, but will still use the GPU if one is available (as long as you do not set `force_device=True`).
If you want to select which GPU is going to be used if needed, you can run the tests with `init_gpu_device=gpu2` for instance.

Sentient07 replied:
@lamblin Thanks! That helped. I have rebased and updated the PR. Please have a look at travis at your convenience.

nouiz commented on theano/tensor/elemwise.py:
```diff
@@ -492,6 +493,8 @@ class Elemwise(OpenMPOp):
 
     """
 
+    __props__ = ("scalar_op", "inplace_pattern")
```
@abergeron @lamblin should the openmp parameter be in this list? I'm not sure of that. It isn't used in \__hash__ and \__eq__. Can you confirm it is fine to not put it here? This would be equivalent to what we do right now.

Sentient07 replied:
@nouiz @abergeron @lamblin should openmp be in this list ?

nouiz replied:
No. This is a special case. OpenMP do not change the result. So if we have 2 nodes in the graph that have the same inputs and are only different by openmp, we want them to be merged. If we add openmp here, they won't be merged. We don't know which one will be kept, but that is less troublesome then not merging them.

Sentient07 commented on theano/sandbox/cuda/opt.py:
```diff
@@ -1902,7 +1895,7 @@ def local_gpu_downsample_factor_max(node):
         x, = node.inputs
         if (x.owner and isinstance(x.owner.op, HostFromGpu)):
-            gpu_ds = GpuDownsampleFactorMax(node.op.ds, node.op.ignore_border)
+            gpu_ds = GpuDownsampleFactorMax(**node.op._props_dict())
```
@nouiz I guess there's been a mistake here. I'm supposed to change only the GPUOps but this isn't an OP, right ?

Sentient07 replied:
@nouiz do i need to retain this change ?

nouiz replied:
This line create an instance of an op. But the original CPU op have more props that then GPU one. So revert this part.

nouiz replied:
you didn't fix this. Revert this change.

lamblin commented:
Also, Travis cannot run since there are conflicts, so you would need to rebase.

nouiz commented:
Travis is failing. Some of the problems are due to flake8:

/home/travis/build/Theano/Theano/theano/tensor/opt.py:645:1: E303 too many blank lines (3)
/home/travis/build/Theano/Theano/theano/tensor/opt.py:649:1: E302 expected 2 blank lines, found 3

The other one could be due to problem in the master that was fixed yesterday. So maybe fix the flake8 problems and we will see how travis respond. You can run the failing test locally to know if they will pass before waiting for travis:

```
FAIL: test3 (theano.tensor.tests.test_opt.test_local_subtensor_lift)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/build/Theano/Theano/theano/tensor/tests/test_opt.py", line 1964, in test3

    assert prog[2].op == tensor.exp

AssertionError

======================================================================

FAIL: theano.tensor.tests.test_opt.test_log1p

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/miniconda2/envs/pyenv/lib/python3.3/site-packages/nose-1.3.4-py3.3.egg/nose/case.py", line 198, in runTest

    self.test(*self.arg)

  File "/home/travis/build/Theano/Theano/theano/tensor/tests/test_opt.py", line 1541, in test_log1p

    T.log1p, tensor.alloc]

AssertionError
```

nouiz commented on theano/tensor/elemwise.py:
```diff
@@ -125,6 +125,7 @@ class DimShuffle(Op):
     check_input = False
+    __props__ = ("input_broadcastable", "new_order")
 
     def __init__(self, input_broadcastable, new_order, inplace=False):
```
For the full change about Dimshuffle inplace, you should change the default to True. Then if it is false, raise an error. Also, update the docstring of this class. In this file, there is a few places, where we set inplace=False. Just remove that paremeter when instanciating a Dimshuffle.

Sentient07 replied:
Yeah, I guess you've mentioned this before or i understood this from one of your comment, but somehow forgot to make this change. I'll update in the upcoming commit

Sentient07 commented:
@nouiz yeah, I had initially thought of addressing pep8 issues once all the other issues in this PR are addressed. I'll fix the pep8 errors as and when flake8 breaks. But, when i tried running the tests after addressing the changes to be made to the other file (```theano/sandbox/gpuarray/opt.py```) I get the following error 
```GpuArrayException: cannot set while device is active in this process: -1```
    
I had gone through [this thread](https://groups.google.com/forum/#!topic/theano-users/nEr9PqQF880), and downgraded cuda to Cuda3 from cuda4. Still this problem persists for me. I'm using GT 750M as GPU and this is how i'm running the tests, 
``THEANO_FLAGS='init_gpu_device=gpu0, lib.cnmem=0.8'  theano-nose theano/sandbox/gpuarray/tests ```
Could you please help me rectifying where i'm going wrong ?

nouiz commented:
THEANO_FLAGS values should not have a space in it. For now, just forget the gpuarray back-end. The change you are doing don't concern it. Changing cuda version won't help on that.

Sentient07 commented:
@nouiz , There are two failures that result from ```theano/tensors/tests/test_opt.py```. I doubt it is because of inplace optimizations. The error message is, 
``` ======================================================================
FAIL: test_matrix_col (theano.tensor.tests.test_opt.Test_lift_transpose_through_dot)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/Ramana/projects/theano/Theano/theano/tensor/tests/test_opt.py", line 5969, in test_matrix_col
    assert str(g) == sg, (str(g), sg)
AssertionError: ('[dot(InplaceDimShuffle{1,0}(InplaceDimShuffle{0,x}(a)), InplaceDimShuffle{1,0}(b))]', '[dot(InplaceDimShuffle{x,0}(a), InplaceDimShuffle{1,0}(b))]')

======================================================================
FAIL: test_row_matrix (theano.tensor.tests.test_opt.Test_lift_transpose_through_dot)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/Ramana/projects/theano/Theano/theano/tensor/tests/test_opt.py", line 5957, in test_row_matrix
    assert str(g) == sg, (str(g), sg)
AssertionError: ('[dot(InplaceDimShuffle{1,0}(b), InplaceDimShuffle{1,0}(InplaceDimShuffle{x,0}(a)))]', '[dot(InplaceDimShuffle{1,0}(b), InplaceDimShuffle{0,x}(a))]')
``` 
Apart from this two failures, I guess all the other tests are good. I had also fixed the pep8 and I hope i've addressed all the comments. Please review at your convenience. Thanks

lamblin commented:
So it looks like there is an optimization simplifying chains of DimShuffle operations that is not applied anymore, I don't know why.

Sentient07 commented:
@nouiz ping

nouiz commented:
Yyou can use the flag optimizer_verbose=True and run it before this PR and
after this PR on the test that fail and compare them. This should tell you
which opt didn't apply in the new code.

On Mon, Mar 14, 2016 at 6:47 PM, Ramana Subramanyam <
notifications@github.com> wrote:

> @nouiz <https://github.com/nouiz> ping
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/4039#issuecomment-196553539>.
>

Sentient07 commented:
@nouiz, i don't know how I missed your comment. Just came to this issue as i was referencing it in another issue.
 As for your question, `local_lift_transpose_through_dot` didn't fit into the new code. I am actually surprised as the failure isn't related to the change made by me. From what i could see, the transpose operation on x has failed to apply.

`    dot(x,y).T -> dot(y.T, x.T)
`

Sentient07 commented:
@nouiz , I apologize for this. I had by mistakenly deleted the optimization for ```local_dimshuffle_lift``` while deleting ```dimshuffle_as_view``` opt. Now it all makes sense. Now I understand how it happened. I backtracked this entire procedure and found the optimization for dimshuffle through elemwise operation and found it missing. The travis pass now.

Sentient07 commented:
@nouiz there is still a test that fails and isn't related to what i've added. Could you please have a look ?

Sentient07 commented:
> This is related to your changes. You changed DImShuffle to be views by default. This broke the tests as it expected them to not be views by default. Change the ground truth to have the DimShuffle inplace/view by default.

I remember changing some DimShuffle to InPlace. But there were few which I was not expecting to be `inplace` or `views` but they were.

Sentient07 commented:
I've fixed the failing tests by replacing `DimShuffle` with `InplaceDimshuffle` and also fixed the pep8 tests. Is it possible to have jenkins test this PR ? :)

nouiz commented:
jenkins test this

nouiz commented on theano/sandbox/cuda/basic_ops.py:
```diff
@@ -185,6 +178,8 @@ class GpuElemwise(GpuOp):
 
     """
 
+    __props__ = ("scalar_op", "inplace_pattern", "sync", )
```
remove and re-add \__eq__. props don't support the dict in inplace_pattern

Sentient07 replied:
Didn't notice the type

nouiz commented on theano/tensor/elemwise.py:
```diff
@@ -504,7 +504,7 @@ class Elemwise(OpenMPOp):
     """
 
-    __props__ = ("scalar_op",)
+    __props__ = ("scalar_op", "inplace_pattern")
```
remove. props don't support dict.

nouiz commented on theano/sandbox/cuda/opt.py:
```diff
@@ -1128,11 +1128,9 @@ def local_gpu_advanced_incsubtensor1(node):
 
-                gpu_op = GpuAdvancedIncSubtensor1(
-                    set_instead_of_inc=set_instead_of_inc)
+                gpu_op = tensor.AdvancedIncSubtensor1(**node.op._props_dict())
```
don't use the CPU op here. Keep the GPU op as before.

Sentient07 replied:
I remember why. Since we're passing the `_props_dict` and this class( `GpuAdvancedIncSubtensor1`) as well it's parents don't have `__init__` or `__props__`, we'd get errors when we do this.

nouiz replied:
Don't use the CPU op and use props. The parent class AdvancedIncSubtensor1 have \__props__ and init. I don't know where you got the information you wrote. I double checked and I have the inverse information.

Sentient07 replied:
This explains the misunderstanding i had in previous comment. If i add `__props__` to the `GpuAdvancedIncSubtensor1` then the `_props_dict()` will use the correct `__props__`

nouiz commented on theano/sandbox/cuda/opt.py:
```diff
@@ -1896,7 +1891,7 @@ def local_gpu_downsample_factor_max(node):
         x, = node.inputs
         if (x.owner and isinstance(x.owner.op, HostFromGpu)):
-            gpu_ds = GpuDownsampleFactorMax(**node.op._props_dict())
+            gpu_ds = GpuDownsampleFactorMax(node.op.ds, node.op.ignore_border)
```
This will fail.

Sentient07 replied:
`GpuDownsampleFactorMax` needs to have `__props__`. After that, passing `_props_dict()` would fix I suppose ?

nouiz replied:
No. GpuDownsampleFactorMax don't implement all the option/props of Pool. This is why we can't use props here. But we check in line 1888 that we didn't included new props in Pool that this opt don't know about.

nouiz commented on theano/tensor/elemwise.py:
```diff
@@ -136,7 +135,7 @@ class DimShuffle(Op):
     check_input = False
     __props__ = ("input_broadcastable", "new_order")
 
-    def __init__(self, input_broadcastable, new_order, inplace=False):
```
Add an assert that inplace is True and if not, give a good user error message.

nouiz commented on theano/tensor/opt.py:
```diff
@@ -675,35 +675,6 @@ def local_lift_transpose_through_dot(node):
-                       60,
-                       'fast_run', 'inplace')
-register_canonicalize(local_dimshuffle_lift)
-register_specialize(local_dimshuffle_lift)
```
Re-add the 2 lines above. They where removed by mistake, we need to keep them.

Sentient07 replied:
Haha, now i understand them :)

Sentient07 replied:
Oh, i've added them before and it is there at the HEAD.

nouiz commented on theano/tensor/elemwise.py:
```diff
@@ -133,7 +133,7 @@ class DimShuffle(Op):
     _f16_ok = True
     check_input = False
-    __props__ = ("input_broadcastable", "new_order")
+    __props__ = ("input_broadcastable", "new_order", "inplace")
```
remove inplace from the props when you add the assert.

Sentient07 replied:
There are few tests that use the inplace parameter. Shall I remove there also ?

nouiz replied:
Yes or delete the test if it isn't needed anymore.

nouiz commented on theano/sandbox/cuda/basic_ops.py:
```diff
@@ -309,10 +309,13 @@ class GpuDimShuffle(GpuOp):
-    def __init__(self, input_broadcastable, new_order):
+    __props__ = ("input_broadcastable", "inplace", "new_order")
+
+    def __init__(self, input_broadcastable, inplace, new_order):
```
Remove the inplace parameter. We want to get rid of it and have it always true.

nouiz commented on theano/sandbox/cuda/opt.py:
```diff
@@ -398,6 +398,9 @@ def local_gpu_dimshuffle_0(node):
             # move the add to a GpuAdd
+            if 'inplace' in node.op._props_dict():
+                import pdb
+                pdb.set_trace()
```
Use "git add -p" to never commit those type of stuff.

Sentient07 replied:
Oh! I forgot to revert this before commit :/

nouiz commented on theano/sandbox/cuda/opt.py:
```diff
@@ -1130,7 +1132,7 @@ def local_gpu_advanced_incsubtensor1(node):
                 gpu_op = tensor.AdvancedIncSubtensor1(**node.op._props_dict())
             else:
-                gpu_op = GPUAdvancedIncSubtensor1_dev20(**node.op._props_dict())
+                gpu_op = theano.sandbox.cuda.basic_ops.GPUAdvancedIncSubtensor1_dev20(**node.op._props_dict())
```
this change is useless, so revert, it make the line too long.

Sentient07 replied:
I've commented in the previous diff regarding this. If i just use `GPUAdvancedIncSubtensor1_dev20` , flake8 tests are failing with 
```
/Users/Ramana/projects/theano/Theano/theano/sandbox/cuda/opt.py:1136:26: F821 undefined name 'GPUAdvancedIncSubtensor1_dev20'
```

nouiz replied:
probably a typo line 1173 does it correctly and without pep8 error

Sentient07 replied:
Ah! My bad

Sentient07 commented on theano/sandbox/cuda/basic_ops.py:
```diff
@@ -121,14 +121,7 @@ class GpuFromHost(GpuOp):
-
-    def __str__(self):
-        return 'GpuFromHost'
+    __props__ = ()
```
I didn't know this then. But do I need to have `__props__` for the classes that don't have `__init__` ?

nouiz replied:
You need \__props__ when the class don't have an \__init__. Don't forget the inheritance. Sometimes the parent class would have one.

Sentient07 replied:
So you mean, since we access the `props_dict`, that would call the `__props__` of the parent class when we don't have it here, right? I understand

Sentient07 commented on theano/sandbox/cuda/opt.py:
```diff
@@ -1128,11 +1128,9 @@ def local_gpu_advanced_incsubtensor1(node):
             else:
-                gpu_op = GpuAdvancedIncSubtensor1_dev20(
-                    set_instead_of_inc=set_instead_of_inc)
+                gpu_op = GPUAdvancedIncSubtensor1_dev20(**node.op._props_dict())
```
I don't know why, but PEP8 complained if i just used `GPUAdvancedIncSubtensor1_dev20`. That's why i did the import as `theano.sandbox.cuda.basic_ops.GPUAdvancedIncSubtensor1_dev20`

Sentient07 replied:
It was done in the next commit

Sentient07 commented on theano/sandbox/cuda/opt.py:
```diff
@@ -2659,15 +2643,15 @@ def gpu_sparse_block_outer_opt(node):
         inputs = _clear_host_from_gpu(node.inputs)
 
-        return [host_from_gpu(GpuSparseBlockOuter(node.op.inplace)(*inputs))]
+        return [host_from_gpu(GpuSparseBlockOuter()(*inputs))]
```
@nouiz why this change ? It was only Dimshuffle that was made not to use the inplace right ?

nouiz replied:
right. This change is harmless. Normally the node should not be inplace at this step. You can probably use props in this opt.

Sentient07 commented:
I've fixed the error on travis and fixed few more redundancy (removing some methods that aren't required when `__props__` is there as discussed with @lamblin . There are a few diff in this commit that I'm unsure about I've commented at those places

Sentient07 commented on theano/sandbox/cuda/basic_ops.py:
```diff
@@ -320,10 +321,18 @@ class GpuDimShuffle(GpuOp):
     check_broadcast = False
 
-    def __init__(self, input_broadcastable, new_order):
+    __props__ = ("input_broadcastable", "new_order", "inplace")
```
I need to have inplace here because there are some optimizations that uses tensor.DimShuffle 's parameters to GpuDimShuffle. So I have added this parameter, and asserted to be always True

nouiz replied:
Don't add it. Remove it from the CPU  props.

Sentient07 commented on theano/sandbox/cuda/basic_ops.py:
```diff
@@ -578,26 +577,6 @@ def __init__(self, reduce_mask, scalar_op, pre_scalar_op=None):
         if pre_scalar_op:
             assert pre_scalar_op.nin == 1
 
-    def __eq__(self, other):
```
It's having `__props__` so I removed these p=methods

nouiz replied:
Keep \__str__ the default string isn't the same and I prefer the one you removed.

Sentient07 commented on theano/sandbox/cuda/basic_ops.py:
```diff
@@ -3656,8 +3635,6 @@ class GpuAllocEmpty(GpuOp):
 
     """
 
-    __props__ = ()
```
No `__init__` so was redundant

nouiz replied:
Keep it. When there is no init, it mean empty props. empty props isn't the same as no props.

nouiz replied:
add back.

Sentient07 commented on theano/sandbox/cuda/basic_ops.py:
```diff
@@ -3866,7 +3835,6 @@ class CopyOnNegativeStrides(GpuOp):
 
     view_map = {0: [0]}
     check_input = False
-    __props__ = ()
```
No `__init__` so removed

nouiz replied:
add back

nouiz commented on theano/sandbox/cuda/opt.py:
```diff
@@ -1131,11 +1127,9 @@ def local_gpu_advanced_incsubtensor1(node):
 
-                gpu_op = GpuAdvancedIncSubtensor1(
-                    set_instead_of_inc=set_instead_of_inc)
+                gpu_op = GpuAdvancedIncSubtensor1(set_instead_of_inc=set_instead_of_inc)
```
use props.

nouiz commented on theano/sandbox/cuda/opt.py:
```diff
@@ -1131,11 +1127,9 @@ def local_gpu_advanced_incsubtensor1(node):
             else:
-                gpu_op = GpuAdvancedIncSubtensor1_dev20(
-                    set_instead_of_inc=set_instead_of_inc)
+                gpu_op = theano.sandbox.cuda.basic_ops.GPUAdvancedIncSubtensor1_dev20(**node.op._props_dict())
```
don't use the full path. It is used elsewhere and it work. So it was probably a typo in the class name.

Sentient07 commented:
In Travis, I'm getting this error, 

```
======================================================================
FAIL: theano.tests.test_printing.test_scan_debugprint1
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda2/envs/pyenv/lib/python2.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/Sentient07/Theano/theano/tests/test_printing.py", line 349, in test_scan_debugprint1
    assert truth.strip() == out.strip()
AssertionError
======================================================================
FAIL: theano.tests.test_printing.test_scan_debugprint3
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda2/envs/pyenv/lib/python2.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/Sentient07/Theano/theano/tests/test_printing.py", line 532, in test_scan_debugprint3
    assert truth.strip() == out.strip()
AssertionError
======================================================================
FAIL: theano.tests.test_printing.test_scan_debugprint5
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda2/envs/pyenv/lib/python2.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/Sentient07/Theano/theano/tests/test_printing.py", line 743, in test_scan_debugprint5
    assert truth.strip() == out.strip()
AssertionError
----------------------------------------------------------------------
Ran 994 tests in 1913.702s
FAILED (SKIP=86, failures=3)
``` 
I'm not able to figure this out

lamblin commented:
Probably, the way an Op is printed has changed after the removal of an `__str__` method.

Please have a look at the content of `truth` and `out` in the tests, it should be easy to figure out.

Then, you could either:
- add back the __str__ method if it is clearer, or
- change the `truth` to mirror the new printed version.

On Tue, Aug 23, 2016, Ramana Subramanyam wrote:
> In Travis, I'm getting this error, 
> 
> ```
> ======================================================================
> FAIL: theano.tests.test_printing.test_scan_debugprint1
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/travis/miniconda2/envs/pyenv/lib/python2.6/site-packages/nose/case.py", line 197, in runTest
>     self.test(*self.arg)
>   File "/home/travis/build/Sentient07/Theano/theano/tests/test_printing.py", line 349, in test_scan_debugprint1
>     assert truth.strip() == out.strip()
> AssertionError
> ======================================================================
> FAIL: theano.tests.test_printing.test_scan_debugprint3
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/travis/miniconda2/envs/pyenv/lib/python2.6/site-packages/nose/case.py", line 197, in runTest
>     self.test(*self.arg)
>   File "/home/travis/build/Sentient07/Theano/theano/tests/test_printing.py", line 532, in test_scan_debugprint3
>     assert truth.strip() == out.strip()
> AssertionError
> ======================================================================
> FAIL: theano.tests.test_printing.test_scan_debugprint5
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/travis/miniconda2/envs/pyenv/lib/python2.6/site-packages/nose/case.py", line 197, in runTest
>     self.test(*self.arg)
>   File "/home/travis/build/Sentient07/Theano/theano/tests/test_printing.py", line 743, in test_scan_debugprint5
>     assert truth.strip() == out.strip()
> AssertionError
> ----------------------------------------------------------------------
> Ran 994 tests in 1913.702s
> FAILED (SKIP=86, failures=3)
> ``` 
> I'm not able to figure this out 
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/Theano/Theano/pull/4039#issuecomment-241758979

-- 
Pascal

Sentient07 commented:
Fixed all the errors in Travis. @nouiz @lamblin could we have Jenkins test on this now?

lamblin commented:
Jenkins test this please

On Fri, Aug 26, 2016, Ramana Subramanyam wrote:
> Fixed all the errors in Travis. @nouiz @lamblin could we have Jenkins test on this now? 
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/Theano/Theano/pull/4039#issuecomment-242678311

-- 
Pascal

nouiz commented on theano/tests/test_printing.py:
```diff
@@ -740,6 +740,9 @@ def test_scan_debugprint5():
     for truth, out in zip(expected_output.split("\n"), lines):
+        if truth.strip() != out.strip():
+            import pdb
+            pdb.set_trace()
```
remove this debugcode

Sentient07 replied:
oh god! I just thought i removed :(

nouiz commented on theano/sandbox/cuda/opt.py:
```diff
@@ -1904,8 +1904,7 @@ def local_gpu_downsample_factor_max_grad(node):
         if (x.owner and isinstance(x.owner.op, HostFromGpu)):
-            gpu_ds_grad = GpuDownsampleFactorMaxGrad(node.op.ds,
-                                                     node.op.ignore_border)
+            gpu_ds_grad = GpuDownsampleFactorMaxGrad(**node.op._props_dict())
```
idem. The node.op here have more props then the GPU ops. So this will fail. Revert this. The assert at line 1898 will safeguard us.

nouiz commented on theano/sandbox/cuda/basic_ops.py:
```diff
@@ -374,24 +388,12 @@ def make_node(self, input):
-    def __hash__(self):
-        return self._hashval
-
-    def __str__(self):
```
Keep the \__str__ it isn't the default version and this one is more beautiful.

nouiz commented on theano/sandbox/cuda/basic_ops.py:
```diff
@@ -575,14 +577,6 @@ def __init__(self, reduce_mask, scalar_op, pre_scalar_op=None):
-            pre = "pre=%s,red=" % str(self.pre_scalar_op)
-        return "GpuCAReduce{%s%s}{%s}" % (
-            pre, str(self.scalar_op),
-            ','.join(str(i) for i in self.reduce_mask))
```
idem, keep this \__str__ method.

Sentient07 replied:
I think you commented on a particular commit. this and one more comment seem to have been addressed at the top of this branch

Sentient07 commented:
@nouiz @lamblin I'm not able to see Jenkins build. I wonder how the tests passed. There are few failing locally and it is expected to fail there as well, right ? Or is Jenkins only for the new backend ?

lamblin commented:
The Jenkins build was among the checks for commit c31d9f7, the link is http://darjeeling.iro.umontreal.ca:8080/job/Theano%20PR%20multijob/146/
Can you access it?

lamblin commented:
Jenkins test this please

Sentient07 commented:
The travis fails with this, 
```
======================================================================
FAIL: theano.tensor.tests.test_gc.test_gc_never_pickles_temporaries
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda2/envs/pyenv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/Theano/Theano/theano/tensor/tests/test_gc.py", line 90, in test_gc_never_pickles_temporaries
    assert len_pre_f == len_post_f
AssertionError
----------------------------------------------------------------------
Ran 3884 tests in 2214.672s
FAILED (SKIP=775, failures=1)
```
I'm not able to reproduce this error locally(using the same flag) could you please restart travis @nouiz ?

nouiz commented:
You can ignore the error. It isn't from this PR. We have this problem in the master of Theano. I'm already working on it. But I still have no idea what is the problem.

I'll make another review of this PR.

jenkins test this

nouiz commented:
I implemented my comments, it was faster then write them.

nouiz commented:
jenkins test this

nouiz commented:
jenkins test this
